bindings/pyroot/src/Cppyy.cxx

```diff
@@ -74,6 +74,7 @@ namespace {
 class ApplicationStarter {
 public:
    ApplicationStarter() {
+      TROOT::Initialize();
```

Comment:
Why are this needed/helpful?

Reply:
Previously ROOT was initialized here: https://github.com/root-project/root/blob/master/bindings/pyroot/src/RootWrapper.cxx#L184-L196

Unfortunately it has shorter lifetime of the `ApplicationStarter` which requires ROOT to be alive to shutdown. Eg. by calling `ApplicationStarter::~ApplicationStarter` ROOT is destroyed.

Reply:
How does calling TROOT::Initialize in the constructor makes gROOT still alive for ```ApplicationStarter::~ApplicationStarter```?   And is that really a 100% guaranteed or would you be better off testing for TROOT still being alive?

Reply:
Without this change I get:
```
(gdb) bt
#0  TCling::~TCling (this=0x555555c85fd0, __in_chrg=<optimized out>)
    at /home/vvassilev/workspace/sources/root/core/metacling/src/TCling.cxx:1439
#1  0x00007fffeb685ece in TCling::~TCling (this=0x555555c85fd0, __in_chrg=<optimized out>)
    at /home/vvassilev/workspace/sources/root/core/metacling/src/TCling.cxx:1452
#2  0x00007ffff45a3458 in TROOT::~TROOT (this=0x7ffff4bc86c0 <ROOT::Internal::GetROOT1()::alloc>, __in_chrg=<optimized out>)
    at /home/vvassilev/workspace/sources/root/core/base/src/TROOT.cxx:1025
#3  0x00007ffff45a0668 in at_exit_of_TROOT () at /home/vvassilev/workspace/sources/root/core/base/src/TROOT.cxx:297
#4  0x00007ffff7a27041 in __run_exit_handlers (status=0, listp=0x7ffff7dcf718 <__exit_funcs>, 
    run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:108
#5  0x00007ffff7a2713a in __GI_exit (status=<optimized out>) at exit.c:139
#6  0x00005555556a9a2f in Py_Exit ()
#7  0x00005555556a7131 in ?? ()
#8  0x00005555556a6a73 in PyErr_PrintEx ()
#9  0x00005555555baf60 in ?? ()
#10 0x0000555555620de4 in Py_Main ()
#11 0x00007ffff7a05b97 in __libc_start_main (main=0x555555620780 <main>, argc=5, argv=0x7fffffffde38, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffde28) at ../csu/libc-start.c:310
#12 0x000055555562069a in _start ()
(gdb) c
Continuing.

Program received signal SIGSEGV, Segmentation fault.
0x0000555557c2e150 in ?? ()
(gdb) bt
#0  0x0000555557c2e150 in ?? ()
#1  0x00007ffff60564e6 in (anonymous namespace)::ApplicationStarter::~ApplicationStarter (
    this=0x7ffff62ff930 <(anonymous namespace)::_applicationStarter>, __in_chrg=<optimized out>)
    at /home/vvassilev/workspace/sources/root/bindings/pyroot/src/Cppyy.cxx:91
#2  0x00007ffff7a27041 in __run_exit_handlers (status=0, listp=0x7ffff7dcf718 <__exit_funcs>, 
    run_list_atexit=run_list_atexit@entry=true, run_dtors=run_dtors@entry=true) at exit.c:108
#3  0x00007ffff7a2713a in __GI_exit (status=<optimized out>) at exit.c:139
#4  0x00005555556a9a2f in Py_Exit ()
#5  0x00005555556a7131 in ?? ()
#6  0x00005555556a6a73 in PyErr_PrintEx ()
#7  0x00005555555baf60 in ?? ()
#8  0x0000555555620de4 in Py_Main ()
#9  0x00007ffff7a05b97 in __libc_start_main (main=0x555555620780 <main>, argc=5, argv=0x7fffffffde38, init=<optimized out>, 
    fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffde28) at ../csu/libc-start.c:310
#10 0x000055555562069a in _start ()
```

`Py_Exit` is a callback from the python interpreter. It starts shutting down the system and run the atexit handlers. If we initialize ROOT outside of `ApplicationStarter`, the atexit handlers of ROOT are run before the ones from the `ApplicationStarter`. This causes a premature call to `TCling::~TCling`. Then we run the `ApplicationStarter` atexit handlers which depend on ROOT being alive and the rest is history :)

I suspect that https://github.com/root-project/root/blob/master/core/base/src/TROOT.cxx#L2091 gets executed before some other (TApplication?) `atexit` function (https://paste.ubuntu.com/p/VbPWsyw7mw/) . According to https://paste.ubuntu.com/p/n2C7wDS6t2/ we call `atexit` on `TROOT::RegisterModule`.

The problem with the current design is that `TApplication` registers a ROOT-dependent atexit function. This in turn does not check if ROOT was initialized and depending on the order of static init it may register it *before* the `at_exit_of_TROOT`. I may make sense to add an assert there...

Reply:
Fair enough.  It makes sense.  The comment can be improved (maybe) by using:
```
      // Insure ROOT's atexit is executed *after* the atexit that calls ApplicationStarter's destructor,
      // by forcing the ROOT's atexit registration now.
      TROOT::Initialize();
```